### ISSUE TITLE
Latex code embedding in jupyter programmatic HTML output not rendering in quarto

### ISSUE THREAD

**Issue by rawlins** [quarto-dev/quarto-cli#7176]
### Bug description

Jupyter renders latex code that appears in programmatic HTML output, but quarto is not doing this, leading to raw latex/MathJax appearing on rendered pages. By "programmatic", I mean anything involving `_repr_html_()` (or the equivalent IPython mime type). The below example uses `IPython.display.HTML` to accomplish this directly. The `%%html` cell magic also has this problem. I've tried to take a look at how MathJax is being invoked by default but I ran into minified code and got stuck.

### Steps to reproduce

Here's a `qmd` that instantiates this problem, obviously it needs Jupyter to run. I've also given markdown for comparison (this behavior matches Jupyter and is essentially the behavior an html output should have; this also provides a partial workaround for now as markdown outputs do support html).

````qmd
---
jupyter: python3
---

```{python}
from IPython.display import Latex, Markdown, HTML
display(Latex("$x_y$"))
display(Markdown("$x_y$"))
display(HTML("$x_y$"))
```
````

### Expected behavior

Here's a screenshot of how this renders in standard Jupyter:

![Screen Shot 2023-10-08 at 1 28 52 PM](https://github.com/quarto-dev/quarto-cli/assets/6290753/5631c649-7a93-432a-a651-1a929442b5a1)

### Actual behavior

Here's a screenshot of the rendered quarto output:

![Screen Shot 2023-10-08 at 1 29 17 PM](https://github.com/quarto-dev/quarto-cli/assets/6290753/aff50eed-bc5f-4398-b168-0b8a4e51c189)

### Your environment

- Jupyterlab: the screenshot above is from 3.1.18, but this is standard behavior for all versions of Jupyterlab. The quarto env actually has jupyter lab 4.0.6 in case that matters (which does mean that MathJax 3 is in play).
- tested on MacOS 11.7.9
- firefox (117.0.1) / chrome (116.0.5845.187)

### Quarto check output

I wasn't able to figure out how to install a prerelease version in a way that isn't system-wide, sorry; but I did scan the changelog and didn't spot anything that would address this. I believe this is the correct pinned pandoc version for this version of quarto, but I also glanced at the pandoc changelog and didn't find anything. Happy to hear if I've missed it, though!

```
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.1: OK
      Dart Sass version 1.58.3: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.3.450
      Path: /Users/advil/miniforge3/envs/quarto2/bin

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.12.0 (Conda)
      Path: /Users/advil/miniforge3/envs/quarto2/bin/python
      Jupyter: 5.3.2
      Kernels: python3, lambda-notebook-dev, lambda-notebook

(\) Checking Jupyter engine render....0.00s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
0.00s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
[✓] Checking Jupyter engine render....OK

[✓] Checking R installation...........(None)

      Unable to locate an installed version of R.
      Install R from https://cloud.r-project.org/
```

**Comment by rawlins** [comment]
Thanks, this is extremely helpful! Definitely straightforward to come up with a workaround based on this. Just a quick follow up, if I'm getting it right:

> * Using HTML() will mark content as raw HTML and this information will be marked like that for Pandoc.
>     * Using only $ .. $ is not raw HTML but rather Markdown
> * Mathjax used with Quarto is [tex-chtml-full](https://docs.mathjax.org/en/latest/web/components/combined.html#tex-chtml-full-component) with default configuration. Rendering $ .. $ is not one of them.

I realize this may be true in the general case, but it does seem to be a miscompatibility with Jupyter, which renders MathJax using `$ ... $` in raw HTML output. I gave the bug example with a `qmd` file because the issue text suggested it, but it happens when rendering directly from an ipynb and perhaps this is the primary issue.

In case it is helpful, here's a notebook version of the same file; the qmd I pasted was an exact output of `quarto convert` on this file: 
[html_bug2.ipynb.zip](https://github.com/quarto-dev/quarto-cli/files/12846014/html_bug2.ipynb.zip)

**Comment by rawlins** [comment]
(Also, I'm a bit confused about the `asis` thing: the docs you link directly show an `IPython.display` example without it in the fourth code cell...)

**Comment by cderv** [comment]
> (Also, I'm a bit confused about the asis thing: the docs you link directly show an IPython.display example without it in the fourth code cell...)

Oh indeed my bad. It seems we handle correctly the `IPython.display` from jupyter... 🤔 The internal structure of our intermediates .md got me confused. 

> I realize this may be true in the general case, but it does seem to be a miscompatibility with Jupyter, which renders MathJax using $ ... $ in raw HTML output.

When you convert it to .html using **nbconvert** ? When you work on the ipynb in Jupyter itself ? 

I believe this is because of Mathjax configuration in the rendered output or preview tools (Jupyter Lab, IDE, nbconvert template, ....). If Mathjax is configured to render `$..$` then it will show as you expect. 

Personally I open ipynb in VSCODE and it does not show as equation for `$...$` syntax. (while it does for the other syntax).

looking at `jupyter notebook` preview - it does configure `inlineMath: [ ['$','$'], ["\\(","\\)"] ],` hence rendering as you expect. 

As I mentioned, if you add the config, it will work.

**Comment by rawlins** [comment]
So, thanks again for the help! I should add that I think quarto is a great package and I'm pretty excited about converting some of my package documentation to it (it currently all lives as plain ipynb files). I maintain two packages that produce relatively complex diagrammatic output with rich IPython reprs, and I do aim for things to work as-is for various compatibility targets (hopefully including quarto). If needed I can add tips to my documentation of course, but it'd be better (for me at least) if no special quarto adjustments would be needed for users who are using quarto from vanilla jupyter outputs using my packages.

Anyways, sorry to keep following up, but I do think there is at least a second bug here beyond the jupyter delimiter mismatch, once I looked into changing delimiters for a specific diagram I'm currently trying to render. Here are two minimally different `qmd`s (generated from plain ipynb files, so completely unadjusted), one of which renders the latex, one of which doesn't. 

Example 1, does not render latex code:

````qmd
---
jupyter: python3
---

```{python}
display(HTML("\(x_y\)"))
```
````


Example 2, does render:

````qmd
---
jupyter: python3
---

```{python}
from IPython.display import Latex
display(Latex("${}$")) # nb: quarto also does not seem to like \(,\) delimiters in this context...
```

```{python}
display(HTML("\(x_y\)"))
```
````

It looks to me like there's something that's trying to detect whether there is latex code and selectively import (or not) MathJax, but is missing code embedded in HTML outputs even with alternate delimiters (perhaps because of the same pandoc raw html thing indicated above?). I also checked this with `\[`,`\]` and the behavior is the same.

**Comment by cderv** [comment]
> Yes, html rendering with $..$ in HTML outputs works in nbvconvert out of the box (as it does in jupyter lab etc).

I believe nbconvert does have the configuration I mentioned in its template : https://github.com/jupyter/nbconvert/blob/4c452c72d09a5f0b28d7fe7deb8b3f1be97a4a7b/share/templates/base/mathjax.html.j2#L18
They have custom configuration it seems. Quarto is using the default Mathjax. 

>  (perhaps because of the same pandoc raw html thing indicated above?).

Yes - this is because this is not parsed as Math by Pandoc. It is seen as raw HTML that should be included asis in the document, and so no transformed to there math version. If the document contains only this, then no mathjax dependency is inserted inside the document. If you add regular math, then Pandoc will see Math in the document, and so insert the dependency. 

<details>
<summary>Example</summary>

````markdown
---
format: html
jupyter: python3
keep-md: true
---

```{python}
from IPython.display import HTML
display(HTML("\(x_y\)"))
```

$x + y$

````

</details>

> If needed I can add tips to my documentation of course, but it'd be better (for me at least) if no special quarto adjustments would be needed for users who are using quarto from vanilla jupyter outputs using my packages.

Can you share more about your package ? 

FWIW math can be forced using a YAML field `math` under `format: html`
````markdown
---
format: 
  html:
    math: true
jupyter: python3
keep-md: true
---

```{python}
from IPython.display import HTML
display(HTML("\(x_y\)"))
```

````

This should work. You could at least document that for now, but maybe looking at your tool we can see how to make it work with Quarto as it is. 

## Next step 

I believe the why is known here: 

* Using `HTML()` really means to Quarto to consider the content as raw HTML and it won't be a good idea to overcome that. Raw HTML means to Pandoc that [`tex_math_dollars`](https://pandoc.org/MANUAL.html#math) feature is not looked into. (this is a markdown syntax feature). Markdown in HTML content won't work when using raw block. 
* Quarto HTML format works differently from other render because it does use Mathjax default configuration which does not consider `$..$` as an inline math syntax. Tools like Jupyter Lab and nbconvert does configure MathJax specifically for this. 

I don't know if we want to maintain a non default Mathjax configuration. We could consider it maybe if this is a large problem among Python package that outputs to Jupyter. 🤔

**Comment by rawlins** [comment]
Thanks for the followup! I have enough to work with here for now and even without the quarto-specific suggestions I had added a markdown mode to the package in question which solves the problem.

> Can you share more about your package ?

Thanks, sure. One is a package for drawing linguistic syntax trees, [svgling](https://github.com/rawlins/svgling); its primary output is pure svg, which in my testing works fine with quarto. It has a mode that uses html for layout primarily so that mathjax can be used in node labels in the diagrams, and it was converting an example of this mode that prompted the question. This package is widely used because it is used by [nltk](https://www.nltk.org/), but really only the svg mode.

The [other package](https://github.com/rawlins/lambda-notebook) (Jupyter lambda notebook) is more niche, and is used for implementing linguistic semantic theories in Jupyter. It heavily uses html + latex diagrams, so it's where I've spent some time debugging this issue. I will say that compatibility for this case is a pain point across everything that is not core jupyter: as you saw it's broken in vscode (no idea if there are workarounds), it's broken in colab (I have a workaround for this one), and probably more. Even for core jupyter, `ipywidgets.HTMLMath` which is supposed to use this same rendering mode as regular `display.HTML` just suffered a regression in jupyterlab 4. So I'm quite used to the problem.

The other main users of this feature that I'm aware of, much bigger than me, are matplotlib [mathtext](https://matplotlib.org/stable/gallery/text_labels_and_annotations/mathtext_demo.html) (which has other rendering means but I believe will use mathjax in jupyter), and some [sagemath](https://doc.sagemath.org/html/en/tutorial/latex.html) rendering (not sure exactly what, I just see this mentioned in the same github issues I'm looking at).

**Comment by rawlins** [comment]
In case you're curious, here's the svgling package docs now rendered via quarto: [https://rawlins.github.io/svgling/](https://rawlins.github.io/svgling/). 

The latex rendering cases that prompted this question are in [sec 7](https://rawlins.github.io/svgling/manual.html#hybrid-htmlsvg-diagrams) of the manual as well as the last example in the diagram gallery. I ended up sticking with my markdown compat mode, as I wasn't able to get the workarounds in this thread that affect mathjax directly to work (but I didn't spend a lot of time on that and I suspect I was doing something wrong).
